source/common/buffer/watermark_buffer.h

```diff
@@ -16,8 +18,12 @@ namespace Buffer {
 class WatermarkBuffer : public OwnedImpl {
 public:
   WatermarkBuffer(std::function<void()> below_low_watermark,
-                  std::function<void()> above_high_watermark)
-      : below_low_watermark_(below_low_watermark), above_high_watermark_(above_high_watermark) {}
+                  std::function<void()> above_high_watermark,
+                  std::function<void()> above_overflow_watermark)
+      : below_low_watermark_(below_low_watermark), above_high_watermark_(above_high_watermark),
+        above_overflow_watermark_(above_overflow_watermark),
+        overflow_multiplier_(Runtime::LoaderSingleton::get().threadsafeSnapshot()->getInteger(
+            "buffer.overflow.high_watermark_multiplier", 2)) {}
```

Comment:
Is a factor of 2 safe based on the maximum expansion of a request that can happen during proxying due to encoding related issues?

Some cases that come to mind:
1.  H2 compressed headers expanding to H1 headers.  I think this is limited by max header size bytes, which is not related to watermark parameters.
2.  H2 connection flow control windows vs buffer watermarks.
3.  H2 control frames which are not limited by flow control.
4.  H1 to H2 chunk body conversions.  I think an 1-byte H1 chunk can be encoded in either 4 or 6 bytes based on which line termination the sender uses.  The H2 data frame header is 9 bytes, so the H2 encoding for a 1 byte frame is 10 bytes (2.5x larger)

Reply:
5. Asymmetric read/write buffer configs.  If the read buffer size is 10x the write buffer, you are very likely to hit the overflow condition after a large read.

Reply:
Good points...
To add to the first bullet, should we verify that the max header size bytes is less-or-equal to overflow watermark value?
Regarding point 5, IIUC it is also the case where the rate of the producer is much faster than the rate of the consumer. In this case I think it's up to the configuration parameters, but if there's a way to warn that a configuration might be wrong, then this is an example where it should.

Reply:
The max header limit excludes framing overhead.  Also, you can end up writing headers and body in the same stack, you need space for both.
I think your changes effectively make it is unsafe to set buffer high water to anything below around 32kb or possibly, yet this is not enforced anywhere in the config plane.  Note that there's widespread use of high watermarks of exactly 32kb.

Reply:
I think the goal of high watermark is to deal with highly unusual situations where normal watermarks have failed us. Examples are where a number of connection buffers are close to their watermark, but not over, and a large read occurs, adding a bunch of data before watermarking or overload manager can respond. Or when we forget to do watermark flow control push back due to a missing feedback path (e.g. for proxy generated direct responses).

Thus, I'm not super worried about the user having to explicitly size them, they should be sizing the normal watermarks (or knobs implying them) based on an understanding of traffic. Whenever there is compression involved, this might be tricky, e.g. zlib or HPACK. Headers are a special case, in general though compression could be 100x. Presumably the basic watermark limit is set to reflect this.

Setting some fixed constant as the fudge factor then makes sense to provide a zero tune capability here IMHO, it probably doesn't matter as long as it's some reasonable distance from the normal operational limit.